tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ConstSimplifierBase.scala

```diff
@@ -29,49 +29,170 @@ abstract class ConstSimplifierBase {
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Division by zero at ${ex.toString}")
+      } else if (right == 1) {
+        leftEx
+      } else {
+        OperEx(TlaArithOper.div, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    // x / x = 1
+    case ex @ OperEx(TlaArithOper.div, leftEx, rightEx) =>
+      if (leftEx == rightEx) {
+        ValEx(TlaInt(1))(intTag)
+      } else {
+        OperEx(TlaArithOper.div, leftEx, rightEx)(intTag)
+      }
+
+    case ex @ OperEx(TlaArithOper.mod, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Mod by zero at ${ex.toString}")
+      } else {
+        ValEx(TlaInt(left % right))(intTag)
+      }
+
+    // x % 1 = 0
+    case ex @ OperEx(TlaArithOper.mod, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        throw new IllegalArgumentException(s"Mod by zero at ${ex.toString}")
       } else if (right == 1) {
-        ValEx(TlaInt(left))(intTag)
+        ValEx(TlaInt(0))(intTag)
+      } else {
+        OperEx(TlaArithOper.mod, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
+
+    // x % x = 0
+    case OperEx(TlaArithOper.mod, leftEx, rightEx) =>
+      if (leftEx == rightEx) {
+        ValEx(TlaInt(0))(intTag)
       } else {
-        ValEx(TlaInt(left * right))(intTag)
+        OperEx(TlaArithOper.mod, leftEx, rightEx)(intTag)
       }
 
-    case OperEx(TlaArithOper.div, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
-      ValEx(TlaInt(left / right))(intTag)
+    case ex @ OperEx(TlaArithOper.exp, ValEx(TlaInt(base)), ValEx(TlaInt(power))) =>
+      if (power < 0) {
+        throw new IllegalArgumentException(s"Negative power at ${ex.toString}")
+      } else if (!power.isValidInt) {
+        throw new IllegalArgumentException(s"Power of ${power} is bigger than an integer at ${ex.toString}")
+      } else {
+        try {
+          // This can take a long time for big base values i.e. 2147484647 ^ 1100000
+          // Maybe we should consider implementing a timeout
+          ValEx(TlaInt(base.pow(power.toInt)))(intTag)
+        } catch {
+          case _: ArithmeticException =>
+            throw new IllegalArgumentException(s"The result of ${ex.toString} exceedes the limit")
+        }
+      }
 
-    case OperEx(TlaArithOper.mod, ValEx(TlaInt(left)), ValEx(TlaInt(right))) =>
-      ValEx(TlaInt(left % right))(intTag)
+    // x ^ 0 = 1
+    // x ^ 1 = x
+    case OperEx(TlaArithOper.exp, leftEx, ValEx(TlaInt(right))) =>
+      if (right == 0) {
+        ValEx(TlaInt(1))(intTag)
+      } else if (right == 1) {
+        leftEx
+      } else {
+        OperEx(TlaArithOper.exp, leftEx, ValEx(TlaInt(right))(intTag))(intTag)
+      }
 
-    case OperEx(TlaArithOper.exp, ValEx(TlaInt(base)), ValEx(TlaInt(power))) =>
-      if (power.isValidInt) {
-        ValEx(TlaInt(base.pow(power.toInt)))(intTag)
+    // 0 ^ x = 0 (except if x = 0 which will match the previous case)
```

Comment:
Literal `0^0` will match the l156 case and default to the `BigInt` implementation of `0.pow(0)`. We can catch it earlier.